vertx-web/src/main/java/io/vertx/ext/web/impl/RoutingContextImpl.java

```diff
@@ -322,7 +322,8 @@ public void reroute(HttpMethod method, String path) {
 
   private Map<Integer, Handler<Void>> getHeadersEndHandlers() {
     if (headersEndHandlers == null) {
-      headersEndHandlers = new LinkedHashMap<>();
+      // order is important we we should traverse backwards
```

Comment:
the `testWithSessions` covers that change. If this is not in place the cookies will not be sent back properly as expected by the test